Sources/SourceKitLSP/Swift/DocumentSymbols.swift

```diff
@@ -0,0 +1,241 @@
+  private var result: [DocumentSymbol] = []
+
+  private init(snapshot: DocumentSnapshot) {
+    self.snapshot = snapshot
+    super.init(viewMode: .sourceAccurate)
+  }
+
+  /// Designated entry point for `DocumentSymbolFinder`.
+  static func find(in nodes: some Sequence<Syntax>, snapshot: DocumentSnapshot) -> [DocumentSymbol] {
+    let visitor = Self(snapshot: snapshot)
+    for node in nodes {
+      visitor.walk(node)
+    }
+    return visitor.result
+  }
+
+  /// Add a symbol with the given parameters to the `result` array.
+  private func record(
+    node: some SyntaxProtocol,
+    name: String,
+    symbolKind: SymbolKind,
+    range: Range<AbsolutePosition>,
+    selection: Range<AbsolutePosition>
+  ) -> SyntaxVisitorContinueKind {
+    guard let rangeLowerBound = snapshot.position(of: range.lowerBound),
+      let rangeUpperBound = snapshot.position(of: range.upperBound),
+      let selectionLowerBound = snapshot.position(of: selection.lowerBound),
+      let selectionUpperBound = snapshot.position(of: selection.upperBound)
+    else {
+      return .skipChildren
+    }
+
+    let children = DocumentSymbolsFinder.find(in: node.children(viewMode: .sourceAccurate), snapshot: snapshot)
+
+    result.append(
+      DocumentSymbol(
+        name: name,
+        kind: symbolKind,
+        range: rangeLowerBound..<rangeUpperBound,
+        selectionRange: selectionLowerBound..<selectionUpperBound,
+        children: children
+      )
+    )
+    return .skipChildren
+  }
+
+  override func visitAny(_ node: Syntax) -> SyntaxVisitorContinueKind {
+    guard let node = node.asProtocol(NamedDeclSyntax.self) else {
+      return .visitChildren
+    }
+    let symbolKind: SymbolKind? = switch node.kind {
+    case .actorDecl: .class
+    case .associatedTypeDecl: .typeParameter
+    case .classDecl: .class
+    case .enumDecl: .enum
+    case .macroDecl: .function // LSP doesn't have a macro symbol kind. Function is the closest.
+    case .operatorDecl: .operator
+    case .precedenceGroupDecl: .operator // LSP doesn't have a precedence group symbol kind. Operator is the closest.
+    case .protocolDecl: .interface
+    case .structDecl: .struct
+    case .typeAliasDecl: .typeParameter // LSP doesn't have a typealias symbol kind. Type parameter is the closest.
+    default: nil
+    }
+
+    guard let symbolKind else {
+      return .visitChildren
+    }
+    return record(
+      node: node,
+      name: node.name.text,
+      symbolKind: symbolKind,
+      range: node.rangeWithoutTrivia,
+      selection: node.name.rangeWithoutTrivia
+    )
+  }
+
+  override func visit(_ node: EnumCaseElementSyntax) -> SyntaxVisitorContinueKind {
+    let rangeEnd =
+      if let parameterClause = node.parameterClause {
+        parameterClause.endPositionBeforeTrailingTrivia
+      } else {
+        node.name.endPositionBeforeTrailingTrivia
+      }
+
+    return record(
+      node: node,
+      name: node.qualifiedDeclName,
```

Comment:
`qualified` to me suggests eg. `Type.member` rather than just the full name (including parentheses).

Reply:
Ah, yeah, qualified really isn’t the term here. Renaming to `declName` in https://github.com/apple/sourcekit-lsp/pull/944